src/os/bluestore/ZonedAllocator.cc

```diff
@@ -145,25 +166,73 @@ void ZonedAllocator::init_rm_free(uint64_t offset, uint64_t length) {
   ceph_assert(length % zone_size == 0);
 
   for ( ; length; length -= zone_size) {
-    advance_write_pointer(++zone_num, zone_size);
+    increment_write_pointer(++zone_num, zone_size);
     ldout(cct, 40) << __func__ << " set zone 0x" << std::hex
 		   << zone_num << " write pointer to 0x" << zone_size << dendl;
   }
 }
 
 bool ZonedAllocator::zoned_get_zones_to_clean(std::deque<uint64_t> *zones_to_clean) {
+  ldout(cct, 10) << __func__ << dendl;
+
+  uint64_t conventional_size = first_seq_zone_num * zone_size;
+  uint64_t sequential_size = size - conventional_size;
+  uint64_t sequential_num_free = num_free - conventional_size;
+  double free_ratio = static_cast<double>(sequential_num_free) / sequential_size;
+
+  ldout(cct, 10) << __func__ << " free size " << sequential_num_free
+		 << " total size " << sequential_size
+		 << " free ratio is " << free_ratio << dendl;
+
   // TODO: make 0.25 tunable
-  if (static_cast<double>(num_free) / size > 0.25) {
+  if (free_ratio > 0.25) {
+    ldout(cct, 10) << " no need to clean" << dendl;
     return false;
   }
   {
     std::lock_guard l(lock);
-    // TODO: populate |zones_to_clean| with the numbers of zones that should be
-    // cleaned.
+
+    if (!cleaning_in_progress_zones.empty()) {
+      ldout(cct, 40) << " cleaning in progress" << dendl;
+      return false;
+    }
+
+    // TODO: make this tunable
+    uint64_t num_zones_to_clean_at_once = 1;
+
+    std::vector<uint64_t> idx(num_zones);
+    std::iota(idx.begin(), idx.end(), 0);
+  
+    if (cct->_conf->subsys.should_gather<ceph_subsys_bluestore, 40>()) {
+      for (size_t i = 0; i < zone_states.size(); ++i) {
+	dout(40) << __func__ << " zone " << i << zone_states[i] << dendl;
+      }
+    }
+
+    std::partial_sort(idx.begin(), idx.begin() + num_zones_to_clean_at_once, idx.end(),
+		      [this](uint64_t i1, uint64_t i2) {
+			return zone_states[i1].num_dead_bytes > zone_states[i2].num_dead_bytes;
+		      });
+
+    ldout(cct, 10) << __func__ << " the zone that needs cleaning is "
+		   << *idx.begin() << " num_dead_bytes = " << zone_states[*idx.begin()].num_dead_bytes
+		   << dendl;
+
+    cleaning_in_progress_zones = {idx.begin(), idx.begin() + num_zones_to_clean_at_once};
+    *zones_to_clean = {idx.begin(), idx.begin() + num_zones_to_clean_at_once};
   }
   return true;
 }
 
+void ZonedAllocator::zoned_mark_zone_clean(uint64_t zone_num) {
```

Comment:
This can and should be done in a bulk manner. Either by passing the list of zones to indicate completions for or (if following my idea about simplification all the cleaning stuff, see another comment) by just calling back from cleaning thread with no parameters which indicates that everything from cleaning_in_progress_zones to be finally cleaned.

Reply:
Done.